app/scripts/migrations/144.ts

```diff
@@ -0,0 +1,301 @@
+  );
+
+  if (typeof currentChainId !== 'string') {
+    global.sentry?.captureException?.(
+      new Error(
+        'Skipping migration: The chain ID resolved from the `NetworkController.selectedNetworkClientId` value is not a string.',
+      ),
+    );
+
+    return state;
+  }
+
+  const updatedSubjects: string[] = [];
+
+  // Add permission to use the current globally selected network to all Snaps
+  // that have the `endowment:ethereum-provider` permission.
+  const entries = Object.entries(permissionControllerState.subjects) as [
+    string,
+    GenericPermissionControllerSubject,
+  ][];
+
+  permissionControllerState.subjects = entries.reduce<
+    PermissionControllerSubjects<PermissionConstraint>
+  >((accumulator, [key, subject]) => {
+    const permissionKeys = Object.keys(subject.permissions);
+    const needsMigration = permissionKeys.includes(
+      SnapEndowments.EthereumProvider,
+    );
+
+    if (!needsMigration) {
+      return {
+        ...accumulator,
+        [key]: subject,
+      };
+    }
+
+    updatedSubjects.push(key);
+
+    const caveatValue = addPermittedEthChainId(
+      getCaveat(subject),
+      currentChainId as `0x${string}`,
+    );
+
+    const newSubject: GenericPermissionControllerSubject = {
+      ...subject,
+      permissions: {
+        ...subject.permissions,
+        [Caip25EndowmentPermissionName]: {
+          caveats: [
+            {
+              type: Caip25CaveatType,
+              value: caveatValue,
+            },
+          ],
+          date: Date.now(),
+          id: nanoid(),
+          invoker: key,
+          parentCapability: Caip25EndowmentPermissionName,
+        },
+      },
+    };
+
+    return {
+      ...accumulator,
+      [key]: newSubject,
+    };
+  }, {});
+
+  const currentDomains = selectedNetworkControllerState.domains;
+  const domains = Object.fromEntries(
+    updatedSubjects.map((subject) => [subject, currentChainId]),
+  );
+
+  selectedNetworkControllerState.domains = {
+    ...currentDomains,
+    ...domains,
+  };
+
+  return state;
+}
+
+/**
+ * Get the existing `endowment:caip25` caveat value for a given subject, if any.
+ * If the caveat does not exist, a new empty caveat value is returned.
+ *
+ * This function does not validate the structure of the existing caveat value,
+ * and assumes that the existing value is valid.
+ *
+ * @param subject - The subject to get the existing caveat value for.
+ * @returns The existing `endowment:caip25` caveat value.
+ */
+function getCaveat(
+  subject: GenericPermissionControllerSubject,
+): Caip25CaveatValue {
+  const existingCaveat = subject.permissions[
+    Caip25EndowmentPermissionName
+  ]?.caveats?.find((caveat) => caveat.type === Caip25CaveatType);
+
+  if (!existingCaveat) {
+    return {
+      isMultichainOrigin: false,
+      optionalScopes: {},
+      requiredScopes: {},
+    };
+  }
+
+  return existingCaveat.value as Caip25CaveatValue;
```

Comment:
Could we remove this type cast?

Reply:
If there's a type guard we could remove it. The caveat is a `CaveatConstraint`, TypeScript can't infer that it's a CAIP-25 permission here.